[WASM] load the sysimage from the executable

vchuravy opened the pull request:
Allows linking the sysimage statically with a binary instead of loading
it through a shared library, currently does not support
multi-versioning.

Keno commented:
So there's two related, but separate changes here:
1. Be able to load a system image from the executable
2. Be able to load a system image that doesn't use multiversioning (yet)

Right now they're combined in this PR. I would suggest we make the load_sysimg_static able to also handle a multi-versioned system image and then add a test (maybe just to the embedding test) that makes sure that we can link a regular sysimage into the regular julia executable and get the desired behavior.

Keno commented:
Also cc @KristofferC, since this may be useful for PkgCompilerX

vtjnash commented on src/init.c:
```diff
@@ -649,7 +649,11 @@ void _julia_init(JL_IMAGE_SEARCH rel)
     jl_prep_sanitizers();
     void *stack_lo, *stack_hi;
     jl_init_stack_limits(1, &stack_lo, &stack_hi);
+#ifdef _OS_EMSCRIPTEN_
```
This doesn’t belong here. It should go inside jl_load_dynamic_library with the other platforms.

vtjnash commented on src/jitlayers.cpp:
```diff
@@ -134,8 +134,10 @@ void addOptimizationPasses(legacy::PassManagerBase *PM, int opt_level,
             PM->add(createLowerPTLSPass(dump_native));
         }
         PM->add(createLowerSimdLoopPass());        // Annotate loop marked with "loopinfo" as LLVM parallel loop
+#if !defined(_OS_EMSCRIPTEN_)
```
The pass manager should be platform independent (though the pass itself may branch on a target triple). Is multiversioning not sufficiently disabled when we pass in a singleton target (the default)?

Keno replied:
I think it's fine to drop from this patch. Since we dump the bc on the wasm bootstrap path, I don't think we ever reach here anyway. We can revisit if we get native bootstrapping working for wasm.

vtjnash commented on src/staticdata.c:
```diff
@@ -1107,7 +1124,7 @@ static void jl_update_all_fptrs(jl_serializer_state *s)
         uintptr_t val = (uintptr_t)&linfos[i];
         uint32_t offset = load_uint32(&val);
-        linfos[i] = NULL;
+        jl_store_unaligned_ptr(&linfos[i], NULL);
```
This array should be aligned

vtjnash commented on src/staticdata.c:
```diff
@@ -237,6 +241,19 @@ static void jl_load_sysimg_so(void)
+void *jl_sysimg_gvars[] __attribute__((weak));
+size_t jl_system_image_size __attribute__((weak)) = 0;
+
+#include <stdio.h>
```
```suggestion
```

vchuravy committed:
Statically load the sysimg through weak symbols

Allows loading the sysimage through weak symbols instead of
loading it through a shared library. This helpful for systems that don't
really have support for dynamic libraries like WASM.

Co-authored-by: Keno Fischer <keno@juliacomputing.com>
Co-authored-by: Valentin Churavy <vchuravy@mit.edu>

vchuravy committed:
rename jl_tls_offset in sysimg

vchuravy committed:
skip length in sysimg_gvars_offset

vchuravy committed:
factor out loading the sysimg and relocating it

vchuravy committed:
relocate functions for static sysimg

vchuravy commented:
I dropped the support for not doing multiversioning, and instead trying to get that to work properly.

I seem to be having issues with the weak symbols so I haven't been able to sufficiently test this...

Keno commented on src/julia_internal.h:
```diff
@@ -509,6 +509,16 @@ extern const int jl_tls_elf_support;
 void jl_start_threads(void);
 
+// -- staticdata.c -- //
+extern char     *jl_sysimg_gvars_base;
```
Should we be putting these into a struct rather than dlsym'ing them individually?

Keno commented:
> I dropped the support for not doing multiversioning

I've looked into this and we can't support multiversioning this way on wasm. The problem is that the function pointer is not fixed at link time, and the relocation to compute differences between function pointers don't exist (also functions aren't in linear memory). We definitely need to avoid trying to compute differences between function pointers at link time on wasm.

vtjnash commented:
We should probably split that pass into the multiversioning and constant-array optimizations. The values here should just be the difference in `function_index` values, as seen in https://stackoverflow.com/questions/44047603/webkit-wasm-stacktrace-debugging, for example. Though I realize that the wasm compiler might not be able to do that yet.

Keno commented:
> Though I realize that the wasm compiler might not be able to do that yet.

*linker, but yes, this relocation just doesn't exist (it doesn't exist on x86 either, but there we're not in the habit of moving individual functions around in the linker, unless we put them each into a separate section).

vtjnash commented:
> *linker
Right, could also be implemented there. Right now we usually expect to resolve this in the compiler (modulo LLVM deciding to drop the feature flag check for it for COFF in D44485 and causing breakage there, sadly—but we revert that PR)

Keno commented:
My point is the wasm linked treats individual functions as an atom, so the difference in function index is not fixed until link time (or even worse post-linking optimization time).
